stl/src/atomic.cpp

```diff
@@ -4,33 +4,87 @@
+    SRWLOCK _Shared_ptr_lock = SRWLOCK_INIT;
 
-_CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // spin until _Shared_ptr_flag successfully set
-#ifdef _M_ARM
-    while (_InterlockedExchange_acq(&_Shared_ptr_flag, 1)) {
-        __yield();
+#if _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA
+    // SPIN LOCK FOR shared_ptr ATOMIC OPERATIONS
+    long _Shared_ptr_flag;
+
+
+    struct _Srw_functions_table {
+        _STD atomic<decltype(&::AcquireSRWLockExclusive)> _Pfn_AcquireSRWLockExclusive{nullptr};
+        _STD atomic<decltype(&::ReleaseSRWLockExclusive)> _Pfn_ReleaseSRWLockExclusive{nullptr};
+    };
+
+    _Srw_functions_table _Table;
+
+    bool _Acuire_srw_functions() {
+        enum class _Shared_ptr_api_level : char {
+            __has_nothing = false,
+            __has_srwlock = true,
+            __not_set,
+        };
+        static _STD atomic<_Shared_ptr_api_level> _Api_level{_Shared_ptr_api_level::__not_set};
+
+        _Shared_ptr_api_level _Current = _Api_level.load(_STD memory_order_acquire);
+        if (_Current == _Shared_ptr_api_level::__not_set) {
+            _Current          = _Shared_ptr_api_level::__has_nothing;
+            HMODULE _Kernel32 = GetModuleHandleW(L"Kernel32.dll");
+            if (_Kernel32 != nullptr) {
+                auto _Pfn_AcquireSRWLockExclusive = reinterpret_cast<decltype(&::AcquireSRWLockExclusive)>(
+                    GetProcAddress(_Kernel32, "AcquireSRWLockExclusive"));
+                auto _Pfn_ReleaseSRWLockExclusive = reinterpret_cast<decltype(&::ReleaseSRWLockExclusive)>(
+                    GetProcAddress(_Kernel32, "ReleaseSRWLockExclusive"));
+                if (_Pfn_AcquireSRWLockExclusive != nullptr && _Pfn_ReleaseSRWLockExclusive != nullptr) {
+                    _Table._Pfn_AcquireSRWLockExclusive.store(_Pfn_AcquireSRWLockExclusive, _STD memory_order_relaxed);
+                    _Table._Pfn_ReleaseSRWLockExclusive.store(_Pfn_ReleaseSRWLockExclusive, _STD memory_order_relaxed);
+                    _Current = _Shared_ptr_api_level::__has_srwlock;
+                }
+            }
+            _Api_level.store(_Current, _STD memory_order_release);
+        }
+        return reinterpret_cast<bool&>(_Current);
     }
-#else // _M_ARM
-    while (_interlockedbittestandset(&_Shared_ptr_flag, 0)) { // set bit 0
+#endif // ^^^ _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA ^^^
+} // unnamed namespace
+
+_EXTERN_C
+
+_CRTIMP2_PURE void __cdecl _Lock_shared_ptr_spin_lock() { // spin until _Shared_ptr_flag successfully set
+#if _STL_WIN32_WINNT >= _STL_WIN32_WINNT_VISTA
+    AcquireSRWLockExclusive(_Shared_ptr_lock);
+#else // ^^^ _STL_WIN32_WINNT >= _STL_WIN32_WINNT_VISTA / _STL_WIN32_WINNT < _STL_WIN32_WINNT_VISTA vvv
+    if (_Acuire_srw_functions()) {
+        _Table._Pfn_AcquireSRWLockExclusive.load(_STD memory_order_relaxed)(&_Shared_ptr_lock);
+    } else {
+        while (_interlockedbittestandset(&_Shared_ptr_flag, 0) != 0) { // set bit 0

```

Comment:
It seems like this should use `cmpxchg`. But also might be unnecessary :)

Reply:
We need to keep the bit 1, and query if it was 0. `lock cmpxchg`, `xchg` or `lock bts` will all do. I'm trying to avoid `lock cmpxchg` overhead, so keeping `lock bts`.

Inner load loop is to avoid needed cache line in exclusive state when we keep waiting for a good value. It is an optimization. note that `lock cmpxchg` will not perform this optimization, even if it does not do store on failure (which is also questionable)

Reply:
> We need to keep the bit 1, and query if it was 0. `lock cmpxchg`, `xchg` or `lock bts` will all do. I'm trying to avoid `lock cmpxchg` overhead, so keeping `lock bts`.
> 

As far as I understand the expensive part of all `lock` instructions is the `lock` part of the instruction, not what the particular instruction does. Why is `lock bts` lower overhead than `lock cmpxchg`? The biggest thing I'm aware of there is that `bts` only works with 32 bit input but `cmpxchg` works with any size input.

> Inner load loop is to avoid needed cache line in exclusive state when we keep waiting for a good value. It is an optimization. note that `lock cmpxchg` will not perform this optimization, even if it does not do store on failure (which is also questionable)

Yes that was always fine and wasn't commenting on that :)

Reply:
I note that the examples in Intel's optimization manuals all use `cmpxchg` or `xchg`.

Reply:
I think apart from the `lock` part, there's a small part of instruction itself (latency, dependencies, regs clobering), and (pre-existing) `lock bts` had it smaller.

Reply:
I see, will investigate that

Reply:
Observed effect from changing one instruction to another is negligible compared to `_mm_pause` and `load` effect, and is less than the variation between runs. Apparently there's no performance reason to prefer one of instruction over another. Picking `xchg` for convenience.